src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -76,11 +76,39 @@ void CoreWorkerDirectTaskSubmitter::OnWorkerIdle(
     // because this is the only place tasks are removed from it.
     if (queue_entry->second.empty()) {
       task_queues_.erase(queue_entry);
+      RAY_LOG(DEBUG) << "Task queue empty, canceling lease request";
+      CancelWorkerLeaseIfNeeded(scheduling_key);
     }
   }
   RequestNewWorkerIfNeeded(scheduling_key);
 }
 
+void CoreWorkerDirectTaskSubmitter::CancelWorkerLeaseIfNeeded(
+    const SchedulingKey &scheduling_key) {
+  auto queue_entry = task_queues_.find(scheduling_key);
+  if (queue_entry != task_queues_.end()) {
+    // There are still pending tasks, so let the worker lease request succeed.
+    return;
+  }
+
+  auto it = pending_lease_requests_.find(scheduling_key);
+  if (it != pending_lease_requests_.end()) {
+    auto &lease_client = it->second.first;
+    auto &lease_id = it->second.second;
+    RAY_LOG(DEBUG) << "Canceling lease request " << lease_id;
+    RAY_UNUSED(lease_client->CancelWorkerLease(
+        lease_id, [this, lease_id, scheduling_key](
+                      const Status &status, const rpc::CancelWorkerLeaseReply &reply) {
+          absl::MutexLock lock(&mu_);
+          if (!status.ok() || reply.success()) {
+            RequestNewWorkerIfNeeded(scheduling_key);
```

Comment:
Why is it needed to request a new worker?

Reply:
This is needed in case a task gets submitted while we're trying to cancel a task. I can add a test case for this.

Reply:
Oh hmm actually I think it's fine since we retry in the callback for the original lease request. Anyway, the test will cover this scenario.